File: apps/client-macos/Sources/TimeTrack/App/AutoTrackingCoordinator.swift
import Foundation

/// PRD §6.1 — bridges `IdleMonitor` decisions to `TimeTracker` writes and `IdleEvent` records.
/// It owns no capture hardware and no policy check: activation is gated upstream by AppDelegate
/// behind `AckGate` (auto-tracking is a capture path — CLAUDE.md §1). The `WorkspaceObserver`
/// (the system edge) calls `tick`/`markAway`/`resume`; this type forwards them to the monitor.
///
/// All callbacks arrive on the main thread (the observer's timer + NSWorkspace notifications),
/// so `TimeTracker` is only ever touched from the main thread.
final class AutoTrackingCoordinator: IdleMonitorDelegate {
    private let tracker: TimeTracker
    private let buffer: TimeEntryBuffering
    private let monitor: IdleMonitor
    private let currentSelection: () -> TimeTracker.Selection
    private let presentAwayPrompt: (_ minutes: Int, _ resolve: @escaping (AwayResolution) -> Void) -> Void
    private let clock: () -> Date
    private let idGen: (Date) -> String
    private let onIdleThresholdCrossed: (Int) -> Void
    /// Fired after EVERY auto open/close so the UI can re-read `TimeTracker`. Defaulted: this
    /// type is built at two call sites and in a dozen tests, and a required parameter on a shared
    /// client type breaks all of them in the same task.
    private let onTrackingStateChanged: () -> Void

    init(
        tracker: TimeTracker,
        buffer: TimeEntryBuffering,
        thresholdSeconds: Int,
        currentSelection: @escaping () -> TimeTracker.Selection,
        presentAwayPrompt: @escaping (_ minutes: Int, _ resolve: @escaping (AwayResolution) -> Void) -> Void,
        clock: @escaping () -> Date = Date.init,
        idGen: @escaping (Date) -> String = { UUIDv7.generate(now: $0) },
        onIdleThresholdCrossed: @escaping (Int) -> Void = { _ in },
        onTrackingStateChanged: @escaping () -> Void = {}
    ) {
        self.tracker = tracker
        self.buffer = buffer
        self.monitor = IdleMonitor(thresholdSeconds: thresholdSeconds, clock: clock)
        self.currentSelection = currentSelection
        self.presentAwayPrompt = presentAwayPrompt
        self.clock = clock
        self.idGen = idGen
        self.onIdleThresholdCrossed = onIdleThresholdCrossed
        self.onTrackingStateChanged = onTrackingStateChanged
        self.monitor.delegate = self
    }

    var monitorState: IdleMonitor.State { monitor.state }

    /// The auto layer stands down entirely while the employee is in a MANUAL session (design
    /// §4): a manually-started entry is the user's own action and must never be auto-stopped or
    /// bridged. Gating the system-edge forwarders (below) is the single mechanism — the monitor
    /// receives no idle/sleep/resume signals during a manual span, so no away cycle, no
    /// spurious IdleEvent, no stop.
    /// `.paused` is included: pause/resume is a manual-only affordance (resume reopens as
    /// `.manual`), so a paused span is a manual session the auto layer must not clobber —
    /// without this, an away→resume cycle could `resolve` and open an AUTO entry over the
    /// paused state (`TimeTracker.start` only self-guards against a second `.tracking` start).
    private var isManualSessionLive: Bool {
        switch tracker.state {
        case .tracking(_, _, _, .manual), .paused: return true
        default: return false
        }
    }

    // Lifecycle (called by AppDelegate, already behind AckGate).
    func activate() { monitor.activate() }
    func deactivate() { monitor.deactivate() }

    // System-edge forwarders (called by WorkspaceObserver). No-ops during a manual session.
    func tick(idleSeconds: Int) { guard !isManualSessionLive else { return }; monitor.tick(idleSeconds: idleSeconds) }
    func markAway() { guard !isManualSessionLive else { return }; monitor.markAway() }
    func resume() { guard !isManualSessionLive else { return }; monitor.resume() }

    // MARK: IdleMonitorDelegate

    func idleMonitorShouldStartTracking(_ monitor: IdleMonitor) {
        let s = currentSelection()
        tracker.start(projectId: s.projectId, taskId: s.taskId, source: .auto)
        onTrackingStateChanged()
    }

    func idleMonitor(_ monitor: IdleMonitor, shouldStopTrackingAt awayStart: Date) {
        tracker.stop(at: awayStart)
        onTrackingStateChanged()
    }

    func idleMonitor(_ monitor: IdleMonitor, didBecomeAwayForSeconds seconds: Int) {
        let minutes = max(1, Int((Double(seconds) / 60.0).rounded()))
        presentAwayPrompt(minutes) { [weak monitor] action in monitor?.resolve(action) }
    }

    func idleMonitor(_ monitor: IdleMonitor, didResolveAwayFrom awayStart: Date, to resume: Date, keeping: Bool) {
        if keeping {
            let s = currentSelection()
            tracker.recordSpan(start: awayStart, end: resume,
                               projectId: s.projectId, taskId: s.taskId, source: .auto)
        }
        enqueueIdleEvent(from: awayStart, to: resume, action: keeping ? .kept : .discarded)
    }

    func idleMonitor(_ monitor: IdleMonitor, didAbandonAwayFrom awayStart: Date, to lastKnown: Date) {
        enqueueIdleEvent(from: awayStart, to: lastKnown, action: .unresolved)
    }

    func idleMonitorDidCrossIdleThreshold(_ monitor: IdleMonitor, afterSeconds seconds: Int) {
        onIdleThresholdCrossed(seconds)
    }

    private func enqueueIdleEvent(from: Date, to: Date, action: ResolvedAction) {
        IdleEventEnqueuer.enqueue(into: buffer, id: idGen(from), from: from, to: to, action: action)
    }
}

File: apps/client-windows/src/NiftyTimer/App/AutoTrackingCoordinator.cs
using NiftyTimer.Storage;
using NiftyTimer.Tracking;

namespace NiftyTimer.App;

/// <summary>
/// PRD §6.1 — bridges <see cref="IdleMonitor"/> decisions to <see cref="TimeTracker"/> writes and
/// IdleEvent records. It owns no capture hardware and performs no policy check of its own:
/// activation is gated upstream in <see cref="AppDelegate"/> behind
/// <see cref="Policy.AckGate"/>, because auto-tracking watches the person continuously and is a
/// capture path under CLAUDE.md §1. The <see cref="SessionObserver"/> calls
/// <see cref="Tick"/>/<see cref="MarkAway"/>/<see cref="Resume"/>; this type forwards them.
///
/// All callbacks arrive on the UI thread (the observer's <c>DispatcherTimer</c> and its window
/// messages), so <see cref="TimeTracker"/> is only ever touched from the UI thread.
/// </summary>
public sealed class AutoTrackingCoordinator : IIdleMonitorDelegate, ISignalReceiver
{
    private readonly TimeTracker _tracker;
    private readonly ITimeEntryBuffer _buffer;
    private readonly IdleMonitor _monitor;
    private readonly Func<TimeTracker.Selection> _currentSelection;
    private readonly Action<int, Action<AwayResolution>> _presentAwayPrompt;
    private readonly Func<DateTimeOffset, string> _idGen;
    private readonly Action<int> _onIdleThresholdCrossed;
    private readonly Action _onTrackingStateChanged;

    public AutoTrackingCoordinator(
        TimeTracker tracker,
        ITimeEntryBuffer buffer,
        int thresholdSeconds,
        Func<TimeTracker.Selection> currentSelection,
        Action<int, Action<AwayResolution>> presentAwayPrompt,
        Func<DateTimeOffset>? clock = null,
        Func<DateTimeOffset, string>? idGen = null,
        Action<int>? onIdleThresholdCrossed = null,
        Action? onTrackingStateChanged = null)
    {
        _tracker = tracker;
        _buffer = buffer;
        _monitor = new IdleMonitor(thresholdSeconds, clock);
        _currentSelection = currentSelection;
        _presentAwayPrompt = presentAwayPrompt;
        _idGen = idGen ?? (now => UuidV7.Generate(now));
        _onIdleThresholdCrossed = onIdleThresholdCrossed ?? (_ => { });
        _onTrackingStateChanged = onTrackingStateChanged ?? (() => { });
        _monitor.Delegate = this;
    }

    public IdleState MonitorState => _monitor.State;

    /// <summary>
    /// The auto layer stands down entirely while the employee is in a MANUAL session: a manually
    /// started entry is the user's own action and must never be auto-stopped or bridged.
    ///
    /// Gating the system-edge forwarders is the single mechanism — the monitor receives no signals
    /// at all during a manual span, so there is no away cycle, no spurious IdleEvent, and no stop.
    ///
    /// Paused counts as manual: pause/resume is a manual-only affordance (resume reopens as
    /// MANUAL), so a paused span is a manual session the auto layer must not clobber. Without this,
    /// an away→resume cycle could resolve and open an AUTO entry over the paused state —
    /// <see cref="TimeTracker.Start"/> only guards against a second start while already tracking.
    /// </summary>
    private bool IsManualSessionLive => _tracker.State switch
    {
        TrackerState.Tracking { Source: TimeTracker.EntrySource.Manual } => true,
        TrackerState.Paused => true,
        _ => false,
    };

    // Lifecycle — called by AppDelegate, already behind the AckGate.
    public void Activate() => _monitor.Activate();

    public void Deactivate() => _monitor.Deactivate();

    // System-edge forwarders. No-ops during a manual session.
    public void Tick(int idleSeconds)
    {
        if (!IsManualSessionLive)
        {
            _monitor.Tick(idleSeconds);
        }
    }

    public void MarkAway()
    {
        if (!IsManualSessionLive)
        {
            _monitor.MarkAway();
        }
    }

    public void Resume()
    {
        if (!IsManualSessionLive)
        {
            _monitor.Resume();
        }
    }

    // Both transitions notify. AUTO writes straight to the tracker, which nothing else observes,
    // so without this the always-visible indicator (PRD §4.2) reads "idle" for the whole
    // login-to-first-idle AUTO span and stays "tracking" after an auto-stop. Notifying on start
    // alone leaves the icon stuck the other way, which is why both are here and both are tested.
    public void ShouldStartTracking()
    {
        var selection = _currentSelection();
        _tracker.Start(selection.ProjectId, selection.TaskId, source: TimeTracker.EntrySource.Auto);
        _onTrackingStateChanged();
    }

    public void ShouldStopTracking(DateTimeOffset awayStart)
    {
        _tracker.Stop(awayStart);
        _onTrackingStateChanged();
    }

    public void DidBecomeAway(int seconds) =>
        _presentAwayPrompt(AwayMinutes.Of(seconds), action => _monitor.Resolve(action));

    public void DidResolveAway(DateTimeOffset awayStart, DateTimeOffset resume, bool keeping)
    {
        if (keeping)
        {
            var selection = _currentSelection();
            _tracker.RecordSpan(
                awayStart,
                resume,
                selection.ProjectId,
                selection.TaskId,
                TimeTracker.EntrySource.Auto);
        }

        Enqueue(awayStart, resume, keeping ? ResolvedAction.Kept : ResolvedAction.Discarded);
    }

    public void DidAbandonAway(DateTimeOffset awayStart, DateTimeOffset lastKnown) =>
        Enqueue(awayStart, lastKnown, ResolvedAction.Unresolved);

    public void DidCrossIdleThreshold(int seconds) => _onIdleThresholdCrossed(seconds);

    private void Enqueue(DateTimeOffset from, DateTimeOffset to, ResolvedAction action) =>
        IdleEventEnqueuer.Enqueue(_buffer, _idGen(from), from, to, action);
}
